File: soc-server/internal/models/agent.go
// ==============================================================================
// Package models - GORM Model: Agent
// File: internal/models/agent.go
// Mô tả: Bảng `agents` lưu thông tin các Endpoint đã cài EDR Agent.
//         Mỗi Agent kết nối về Server qua gRPC/mTLS với chứng chỉ số riêng.
// ==============================================================================

package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Agent - GORM model cho bảng `agents`
// Mỗi record đại diện cho 1 Endpoint (máy tính) đã cài đặt EDR Agent
type Agent struct {
	// ID - Khóa chính UUID, tự động sinh khi tạo mới
	ID string `gorm:"size:36;primaryKey" json:"id"`

	// Hostname - Tên máy tính của Endpoint (ví dụ: "WORKSTATION-01")
	Hostname string `gorm:"type:varchar(255);not null;index" json:"hostname"`

	// IPAddress - Địa chỉ IP hiện tại của Endpoint
	IPAddress string `gorm:"type:varchar(45);not null" json:"ip_address"`

	// OSType - Hệ điều hành: "windows", "linux", "macos"
	OSType string `gorm:"type:varchar(50);not null" json:"os_type"`

	// Status - Trạng thái hiện tại: "online" hoặc "offline"
	// Được cập nhật dựa trên heartbeat và Redis cache
	Status string `gorm:"type:varchar(20);default:'offline';index" json:"status"`

	// LastHeartbeatAt - Thời điểm nhận heartbeat gần nhất từ Agent
	// Nếu quá lâu không nhận → đánh dấu Agent offline
	LastHeartbeatAt *time.Time `gorm:"type:datetime\(3\)" json:"last_heartbeat_at"`

	// MTLSCertFingerprint - SHA256 fingerprint của chứng chỉ mTLS Client
	// Dùng để xác thực Agent khi kết nối gRPC
	MTLSCertFingerprint *string `gorm:"type:varchar(255);uniqueIndex" json:"mtls_cert_fingerprint"`

	// AgentVersion - Phiên bản Agent đang chạy trên Endpoint
	AgentVersion string `gorm:"type:varchar(50)" json:"agent_version"`

	// Timestamps chuẩn GORM
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deleted_at,omitempty"`
}

// TableName - Chỉ định tên bảng trong PostgreSQL
func (Agent) TableName() string {
	return "agents"
}

// BeforeCreate - GORM Hook: Tự động sinh UUID trước khi INSERT
func (a *Agent) BeforeCreate(tx *gorm.DB) error {
	if a.ID == "" {
		a.ID = uuid.New().String()
	}
	return nil
}

File: soc-server-python/app/models/agent.py
# ==============================================================================
# app/models/agent.py - SQLAlchemy Model: Agent
# Tương đương: internal/models/agent.go
# ==============================================================================

import uuid
from datetime import datetime
from typing import Optional

from sqlalchemy import String, DateTime, Boolean
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Agent(Base):
    __tablename__ = "agents"

    # ID UUID (tự sinh)
    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))

    # Hostname máy endpoint
    hostname: Mapped[str] = mapped_column(String(255), nullable=False, index=True)

    # Địa chỉ IP
    ip_address: Mapped[str] = mapped_column(String(45), nullable=False)

    # Hệ điều hành: windows, linux, macos
    os_type: Mapped[str] = mapped_column(String(50), nullable=False)

    # Trạng thái: online, offline
    status: Mapped[str] = mapped_column(String(20), default="offline", index=True)

    # Thời điểm nhận heartbeat gần nhất
    last_heartbeat_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)

    # SHA256 fingerprint chứng chỉ mTLS
    # Ghi chú: Go GORM đặt tên cột là m_tls_cert_fingerprint (từ MTLSCertFingerprint)
    mtls_cert_fingerprint: Mapped[Optional[str]] = mapped_column(
        String(255), name="m_tls_cert_fingerprint", unique=True, nullable=True
    )

    # Phiên bản Agent
    agent_version: Mapped[str] = mapped_column(String(50), default="")

    # Timestamps
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    deleted_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True, index=True)

    # Relations
    alerts: Mapped[list["Alert"]] = relationship("Alert", back_populates="agent", lazy="select")  # type: ignore
